### PR TITLE
Improve current GitHub Action CI with multi OS support

### LINKED ISSUE #740: python3.9 incompatibility due to onnxruntime
ivadomed doesn't install on python3.9 because onnxruntime is pinned to an old version that 

a. doesn't provide packages for 3.9: https://pypi.org/project/onnxruntime/1.4.0/#files
b. doesn't even provide an `sdist` for anything

```
(venv) [kousu@requiem ivadomed]$ pip install -e .
Obtaining file:///home/kousu/src/neuropoly/ivadomed
Collecting bids-neuropoly>=0.2
  Using cached bids_neuropoly-0.2-py3-none-any.whl (3.1 kB)
Collecting csv-diff>=1.0
  Using cached csv_diff-1.1-py3-none-any.whl (12 kB)
Collecting h5py==2.10.0
  Using cached h5py-2.10.0.tar.gz (301 kB)
Collecting joblib
  Using cached joblib-1.0.1-py3-none-any.whl (303 kB)
Collecting matplotlib>=3.3.0
  Using cached matplotlib-3.4.0-cp39-cp39-manylinux1_x86_64.whl (10.3 MB)
Collecting nibabel
  Using cached nibabel-3.2.1-py3-none-any.whl (3.3 MB)
ERROR: Could not find a version that satisfies the requirement onnxruntime==1.4.0 (from ivadomed)
ERROR: No matching distribution found for onnxruntime==1.4.0
```

Their latest version https://pypi.org/project/onnxruntime/1.7.0/#files does provide 3.9 packages.

### LINKED ISSUE #778: CI Test gets stuck in an infinite loop
## Issue description
testing/functional_tests/test_automate_training.py does not terminate on macOS and Windows CI environments.

### Current behavior

The test gets stuck.

### Expected behavior
 
The test should terminate within minutes.

### Steps to reproduce

https://github.com/ssinad/ivadomed/actions/runs/814749769

### System description

macOS 10.15, windows-latest

### Installed packages

<details open>
<summary> Output of <code>pip freeze</code> </summary>

```

```

</details>

### PR DESCRIPTION
**Pull request by dyt811**
#### GitHub

- [X] I've given this PR a concise, self-descriptive, and meaningful title
- [X] I've linked relevant issues in the PR body
- [X] I've applied [the relevant labels](https://www.neuro.polymtl.ca/software/contributing#pr_labels) to this PR
- [x] I've assigned a reviewer

#### PR contents

- [X] I've consulted [ivadomed's internal developer documentation](https://github.com/ivadomed/ivadomed/wiki) to ensure my contribution is in line with any relevant design decisions
- [X] I've added [relevant tests](https://github.com/ivadomed/ivadomed/wiki/tests) for my contribution
- [x] I've updated the [relevant documentation](https://github.com/ivadomed/ivadomed/wiki/documentation) for my changes, including argparse descriptions, docstrings, and ReadTheDocs tutorial pages

### Why: 
- We currently have no CI for macOS/Windows
- Pip install -r requirement.txt will fail on Windows side because pytorch 1.5.0 does not have wheel on Windows
- We don’t have a Conda script to deploy environment even if we want to. 
- Dev don’t have quick way to deploy Dev environment (that includes pytest etc) via pip or conda 
- User/Dev don’t have quick way to deploy GPU related environment easily. 

### What:

- Current GitHub Action CI does not fully address Windows OR even macOS platform testing due to earlier bug which has been recently resolved. See #604 #755. Now we can properly robustly test Windows integration. I have incorporated many ADS CI tips and tricks learned from there (https://github.com/neuropoly/axondeepseg/pull/479) to improve our existing matrix testing to be more comprehensive and thorough while adding additional comments to make it easier for subsequent future updates. 

### How: 
Applying cascading `requirements.txt` without changing anything major it now allows us to accommodate various deployment scenarios while maintaining single source of truth
- `Requirements_common.txt` shared with `Conda` and `setup.py` cascade to various different higher level requirements. 
- `Conda` now supported via `environment.yml` which refer to the `requirements.txt` to respect SINGLE source of truth
- User deploys `requirements.txt` with CPU support, standard deployment scenario. 
- Advanced user deploys `requirements_gpu.txt` for GPU support. 
- Dev/CI/RTD uses `requirements_dev.txt` which cascades off the main requirements.txt

### How to test 
- Create new venv/conda env and validate using whichever conda/venv/pip installation setup approach you are most comfortable playing with. Report back issues and concerns. 
- `requirements_dev.txt` might require some extra attention as not all sphinx build/wheel are available for Windows? 
- Could also use more feedback on the documentation side

## Linked issues

Resolves #778, also provide CI for its precursor #633 #768

### CODE REVIEW

**Comment by dyt811** [comment]
Took a while to realize that macOS does not support the Wheel +CPU tag... sigh. 
image

No wonder it fails on macOS only. FML.

**Comment by charleygros** [comment]
@dyt811 : Thanks for all this! I'd be happy to review, but I'm scared I don't have a relevant skill set for this. Maybe someone else in the team would be more appropriate? Let me know :-)

**Comment by dyt811** [comment]
Closer examination of the output showed that automated_training_process was skipped accidentally. Once properly restored, it suggested that we are still running into issue with Mac/Windows hanging and it is on top of the earlier mentioned problem with HDF5. Root cause still not fully identified...

File: .github/workflows/run_tests.yml
Comment (ssinad) [inline .github/workflows/run_tests.yml:60]: @dyt811 Which requirements file does this one install?

File: environment.yml
Comment (ssinad) [inline environment.yml:1]: @dyt811  Could you explain what the purpose of this file is?

File: setup.py
Comment (ssinad) [inline setup.py:8]:
```python
requirements = f.readlines()
```
would also work.

**Review by ssinad — CHANGES_REQUESTED** [review]
@dyt811
I would recommend that you get another approval from someone else on the team before you merge.

File: .github/workflows/run_tests.yml
Comment (ssinad) [reply .github/workflows/run_tests.yml:60]: Maybe you could add
```yaml
path: requirements_dev.txt
```

File: requirements_dev.txt
Comment (cakester) [inline requirements_dev.txt:11]: no version number? in the context that you just deleted this has a version number https://github.com/ivadomed/ivadomed/pull/757/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552L19

Comment (dyt811) [reply requirements_dev.txt:11]: Good catch. Yeah, I think this PR started before your PR so this needs to be manually added in this new file. I will added it.

File: .github/workflows/run_tests.yml
Comment (ssinad) [inline .github/workflows/run_tests.yml:66]: @dyt811 Consider using the previous step instead of hardcoding package versions.

File: docs/source/installation.rst
Comment (jcohenadad) [inline docs/source/installation.rst:30]:
```suggestion
There are fundamentally two different approaches to install ``ivadomed``:
```

File: .github/workflows/run_tests.yml
Comment (ssinad) [inline .github/workflows/run_tests.yml:60]:
```suggestion
        uses: py-actions/py-dependency-install@v2
        path: requirements_dev.txt
```

File: setup.py
Comment (ssinad) [inline setup.py:8]:
```suggestion
    requirements = f.readlines()
```

File: .github/workflows/pre-commit.yml
Comment (ssinad) [inline .github/workflows/pre-commit.yml:43]:
```suggestion
              if [[ -f requirements_dev.txt ]]; then pip install -r requirements_dev.txt; fi
```
See [here](https://github.com/ivadomed/ivadomed/wiki/variables#if-statements).

File: .github/workflows/run_tests.yml
Comment (dyt811) [reply .github/workflows/run_tests.yml:66]: Good point!

File: .github/workflows/run_tests.yml
Comment (dyt811) [reply .github/workflows/run_tests.yml:33]: I believe this [disables fast fail in GitHub CI](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference) which result in more likelihood of revealing bugs in CI. I have actually [copied this bit from the template GitHub Action from ADS](https://github.com/neuropoly/axondeepseg/blob/master/.github/workflows/run_tests.yaml#L44).

File: environment.yml
Comment (dyt811) [reply environment.yml:1]: This environment YML is provided such that conda users can easily instantiate a conda environment while still be able to source the configure file from a single source of truth which is shared across multiple dev/user environment setups such as: 1) CPU Dev, 2) GPU users. 

Conda user can certainly just instantiate their own environment and pip install the requirement as well, just save them a few key strokes.

File: setup.py
Comment (dyt811) [reply setup.py:8]: TIL. Cool. Definitely didn't know that before. :)

File: .github/workflows/pre-commit.yml
Comment (dyt811) [reply .github/workflows/pre-commit.yml:43]: COOL. TIL. Thanks for the info!

**Review by ssinad — APPROVED** [review]
Great job. I would still recommend you get another approval before you merge, though.

**Comment by jcohenadad** [comment]
> Great job. I would still recommend you get another approval before you merge, though.

we are a very small team and there are *many* other urgent priorities so unfortunately we need to live with PRs not being reviewed as much as everyone would like--

**Comment by ssinad** [comment]
> > Great job. I would still recommend you get another approval before you merge, though.
> 
> we are a very small team and there are _many_ other urgent priorities so unfortunately we need to live with PRs not being reviewed as much as everyone would like--

I understand.

File: requirements_dev.txt
Comment (ssinad) [inline requirements_dev.txt:6]:
```suggestion
pytest~=6.2
```

File: requirements_dev.txt
Comment (ssinad) [inline requirements_dev.txt:9]:
```suggestion
pytest-ordering~=0.6
```

File: requirements_dev.txt
Comment (dyt811) [reply requirements_dev.txt:6]: Good catch. Yup. Better to version these. Gonna add it in.

File: requirements_dev.txt
Comment (dyt811) [reply requirements_dev.txt:9]: Yup. Will add this in as well. I was REALLY hoping that we will get CI to a state where daily CI runs will catch any subversion induced problems but I think there is a whole bunch of tests to enhance before that happens.

File: .github/workflows/pre-commit.yml
Comment (konantian) [inline .github/workflows/pre-commit.yml:15]: We use "macos-11" here is because "macos-latest" is equals to "macos-10.15" instead of "macos-11".

File: .github/workflows/run_tests.yml
Comment (konantian) [inline .github/workflows/run_tests.yml:42]: We use "macos-11" here is because "macos-latest" is equals to "macos-10.15" instead of "macos-11".

**Review by konantian — COMMENTED** [review]
Looks good to me, I've tested them on both Windows and macOs locally, no issues found.

File: .github/workflows/run_tests.yml
Comment (ssinad) [reply .github/workflows/run_tests.yml:42]: @dyt811 
macOS 11 is not available in GitHub actions yet. That is why the tests are failing. See [here](https://docs.github.com/en/actions/using-github-hosted-runners/about-github-hosted-runners).

> Note: The macOS 11 virtual environment is currently provided as a private preview only. Any users or organizations that are already using this runner can continue using it, but we're not accepting any further users or organizations at this time. The macos-latest YAML workflow label still uses the macOS 10.15 virtual environment.

File: .github/workflows/run_tests.yml
Comment (dyt811) [reply .github/workflows/run_tests.yml:60]: Yup. Change included.

File: setup.py
Comment (dyt811) [reply setup.py]: Updated.

File: .github/workflows/run_tests.yml
Comment (dyt811) [reply .github/workflows/run_tests.yml:42]: Good point. I thought it was available but didn't read the fine print about it being "private preview only". Gonna revert that and merge PR then.

File: environment.yml
Comment (kousu) [inline environment.yml:7]: @dyt811 can you explain why you depend on `anaconda` explicitly?

This doesn't seem to be compatible with https://github.com/conda-forge/miniforge, which is what I picked to use in https://github.com/neuropoly/computers/pull/120/ on the theory that it would be smaller and better supported.

Removing this one line seems to make ivadomed fully compatible with miniforge.

Comment (dyt811) [reply environment.yml:7]: Yup. No need to use anaconda. Miniconda is good enough. 👍

**Pull request merged**
